### DISCUSSION TITLE
Why can't functions run on draft order based checkouts?

### DISCUSSION THREAD

**Discussion by Sam115200** [Shopify/function-examples#393] [Feedback: General Improvements]
This issue is related to #357.

What was reported in #357 is that draft order based checkouts are running Shopify functions recently. But this has been actually happening for at least 5 months or so. This is not a new behavior. I tested draft order based checkouts 5 months ago using functions and they were running fine, the only issue I observed was that the cart line details and delivery groups details were not being reflected in the function input, the cart lines/delivery groups details were always set to be null. Other than that, everything seemed to work fine.

This means that there are a lot of merchants in the last 5 months who set Shopify functions on their store, while they also had some other app which was creating draft order based checkouts on their store (There are a lot of apps who use draft order API). They were able to successfully use the functions to modify the checkout, even when it was generated using draft order API, since functions were running fine on these checkouts as well.

After #357 was resolved, functions are not allowed to run on the draft based checkouts anymore, so all of those merchants who were using functions successfully in the last 5 months (while having other apps which caused their checkout to be draft order based), now have huge problems, since functions were suddenly not running anymore on the checkout and all the logic they set in the function to show correct options at the checkout went down the drain. 

This has caused a lot of problems for those merchants and I've been trying to mitigate it as much as I can by communicating to them and explaining to them, why is this happening. But if the functions team can answer questions mentioned below, that would be very helpful.

1 -  When can functions run on the draft order based checkouts again? what should I tell merchants?
2 - Is there any way to make this feature high priority? 

This is creating a `fragmented experience for merchants` (even for those who are new to functions), they set the function on their store and it works, then they install another app and suddenly their checkout breaks, since functions stop working and they have to go through a lot, just to figure out what happened and how to fix it. 

It is `nice` to add new features into functions, but functions are useless when a lot of merchants can't even use them, since they are already using other apps which use draft order API. Enabling this `essential` feature will create a more smoother experience for merchants and reduce a lot of issues.

**Comment by thosakwe** [comment]
This would be a great addition, and it would help with overall app compatibility.

**Comment by cupnoodle** [comment]
Agreed on the fragmentation part, there’s quite some merchants using functions were confused because the functions were not running sometimes (on draft order checkout), please look into this or at least make an option for merchant or developer to run functions on draft order.

**Comment by thosakwe** [comment]
Automatic discount functions work in draft orders now, but I'm not sure about other Functions APIs.

**Reply by Sam115200** [reply]
Other functions still do not support draft order checkout at all.

**Reply by mubbo** [reply]
I think its more that shopify hasnt yet moved the checkout that draft orders use to the "new" checkout. The url structure of the checkout for draft orders is completely different which leads me to believe its two separate checkouts. Im guessing the old one which didnt have support for the new function apis built in or at least all of them.

**Comment by mufftee** [comment]
We also need Functions in Draft orders because we get a ton of uninstalls if the people use Hulk Options for example. These "Customiziation Apps" often work with draft orders.

**Reply by Jayanthi-naik** [reply]
I am trying to disable discounts application on draft orders using shopify functions. Is it possible?

**Comment by nickwesselman** [comment]
Moving to General as this affects multiple APIs.

**Reply by mufftee** [reply]
Thank god!

**Comment by Sam115200** [comment]
Bump for this, as this is causing huge problems and really bad and fragmented experience for merchants.

**Comment by mubbo** [comment]
Wow, this is a huge oversight. Honestly why is there even a separate checkout to begin with.  

I have plus Merchants who wish to allow certain customers who pass a credit check to buy on invoice. The results of the credit check are stored in a metafield attached to the customer.  When the customer goes to the checkout and they have the appropriate score they can purchase on invoice, if not the payment method is hidden.

Now one merchant has an app that applies discounts to the order, but it creates a draft order and routes the customer to the draft order link instead of going to to the normal checkout. and suddenly for customers who shouldnt be able to purchase on invoice the option is available.  This could cause serious headaches for the merchant having to deal with this huge potential for fraud.

In any of the other APIs I would say ok, it takes a bit of time to roll things out everywhere, but the payment API that just needs to work and work everywhere.

Cheers,

Gary

**Comment by Oteiza-a** [comment]
I'll bump this too! It's quite frustrating to tell merchants that there is no way that the customizations will work for them for a reason totally out of our control. Around 30% of the merchants that use my customizations have problems with this. I hope Shopify will support draft orders with functions 😪

**Comment by cm-toshio-seki** [comment]
Will the payment customization API work for draft orders?

I would like to disable postpaid payment if a product with a specific tag exists in an order created by draft order using the payment customization API of shopfiy functions.

**Reply by mubbo** [reply]
Unless they changed something the answer is no. Apparently draft orders are run through a different checkout which does not yet support functions. I also want to disable a method based on certain criteria and it does not work for draft orders.

**Reply by cm-toshio-seki** [reply]
Thank you for your response. I hope that Shopify functions will support draft orders too.

**Comment by christianoneillshopify** [comment]
Hi Team! Do you know of any apps or ways to modify the 'draft checkout' as apps such as 'Hidepay' don't wort on the 'draft checkout' leaving merchants very confused. (See photos attached)
Any help is appreciated.
Many thanks,
Shopify Payments OFF
<img width="294" alt="Checkout When SP OFF" src="https://github.com/Shopify/function-examples/assets/157732568/86dc9d5f-d26a-48fa-ab50-3d2e061ac1a9">
Shopify Payments ON
<img width="288" alt="Checkout When SP ON" src="https://github.com/Shopify/function-examples/assets/157732568/785af8aa-d1c8-4034-afcf-617c10a89f3e">

**Comment by 50bbx** [comment]
According to this timeline there is no mention whatsoever to Draft Orders in future Shopify Functions features: https://shopify.dev/docs/apps/checkout

**Comment by tomhooker** [comment]
This is really bad over sight in my opinion. This has ruined the customer experience for some of my clients, as we just assumed our functions would be working, a lot of merchants use draft order checkouts!

**Reply by tomhooker** [reply]
Can I also point how still they do not list the draft orders as a limitation. Some serious oversight on Shopifys end. https://shopify.dev/docs/apps/functions#limitations-and-considerations

**Comment by hterhanian** [comment]
Please work on this! Shopify sales talked up it's B2B capabilities when selling me on the Plus subscription and I keep running into missing functionality with draft orders which I absolutely need when working with wholesale buyers.

**Comment by 50bbx** [comment]
Does the first video on the [Summer '24 page](https://www.shopify.com/editions/summer2024) talk about this? It looks like Draft Order now works with payment customisations but then if you search on the page you'll find no information about this.

**Reply by Oteiza-a** [reply]
It still doesn't work in draft orders, for me at least. Just tested it.

**Reply by christianoneillshopify** [reply]
Any news on when apps will work on draft checkout?

**Reply by kylebuildsstuff** [reply]
Just wanted to confirm if product discount functions are supposed to be working on draft order checkouts now? 

I know there was news about compatibility with UI extension but it still doesn't seem to work for my product discount function. 

Edit: nvm product discount is applying on draft checkouts

**Comment by mdriess** [comment]
This is also not great for POS, since emailed carts use draft order checkout now 😅

**Comment by gabrielvictoria96** [comment]
Any updates on this? Is order discount supposed to work on draft orders? Because it doesn't work for me =(

**Comment by abmenzel** [comment]
This is a big issue for us, and I don't understand why it's not even mentioned on the roadmap https://shopify.dev/docs/apps/build/checkout#product-roadmap

Any updates?

**Comment by kalenjordan** [comment]
Also need this.

**Comment by smiddygroup** [comment]
Can we get a timeline on this? @shopify-admins

**Comment by smshahriar10** [comment]
Has the Function API started working with draft orders? I just tested the Validation API, and it’s suddenly working with draft orders.

<img width="742" alt="draft-orders" src="https://github.com/user-attachments/assets/39397df0-b790-45b5-97f7-c90bca5383a0">

**Reply by mdriess** [reply]
Just checked; we are also seeing them work in draft orders. UI extensions too.

**Reply by patjakubik** [reply]
Same here.

**Reply by cupnoodle** [reply]
Was it just validation API? I noticed payment customization API still does not work on draft order

**Reply by mdriess** [reply]
(We haven't noticed / checked anything around payment customizations on our side)

**Reply by smshahriar10** [reply]
Looks like Validation API stopped working again for the draft orders today.

**Comment by Oteiza-a** [comment]
I can confirm Payment Customizations are still not working with Draft Orders. 

@shopify-admins Is there a possibility that this is added to the roadmap? It would make Payment customizations a more complete product.

**Comment by gil--** [comment]
Closing as draft orders now support payment & delivery functions: https://changelog.shopify.com/posts/draft-orders-now-support-delivery-and-payment-method-customizations-and-checkout-rules

**Comment by smshahriar10** [comment]
I tested today, and it seems the Validation Function API isn’t working for draft orders. Does this mean only payment and delivery functions currently support draft orders? @gil--

**Reply by gil--** [reply]
Validation Functions are compatible with draft orders.

**Comment by AlexHooperDev** [comment]
Is there an input available for checkout functions that indicates whether the checkout is in use for a draft order? (or even, better, an invoice checkout that the customer has been sent)?

**Reply by gil--** [reply]
We do not surface the channel or origination of an order but draft orders originating from the admin can manually bypass validation rules:

<img width="634" alt="Screenshot 2025-03-13 at 2 53 39 PM" src="https://github.com/user-attachments/assets/2cc6b4e4-bd09-4440-93ee-98e6480e8a57" />

**Reply by Oteiza-a** [reply]
@gil--  What does it mean that I don't see the "Checkout rules option" when sharing the checkout link? 
<img width="636" alt="Screenshot 2025-03-26 at 12 37 07 PM" src="https://github.com/user-attachments/assets/d32d37d5-0e5b-4bd1-9631-ed1e1bb00036" />

I'm in a weird situation where a merchant reports that draft order checkouts created from the POS (for the customer to pay in the online checkout) do not work with payment customizations. The POS draft orders are redirected from a link in this format -> https://[shop url]/29295648/daco/ee6e13bbe7ec4c7100bf122547846972 and end up in a link in this format: https://[shop url]/checkouts/do/275c7f717e7fd74890a2cb3cf63ccb3c

But if we open the draft order with the link from clicking "Share" button for that draft order, it works perfectly (the payment customizations are applied).

**Reply by gil--** [reply]
Let me look into the POS question but if you do not have any active Checkout Rules (Validation functions) on the store, then you won't see Checkout rules as an option/setting.

**Reply by Oteiza-a** [reply]
Thank you @gil-- for the clarification about the checkout rules. Regarding the POS draft orders, the merchant reports that it used to work but now it doesn't, though, he may be mistaken.

**Reply by Jrschellenberg** [reply]
@gil--  Is there any plan on adding in an input variable setting so the function can detect if it is a Draft order?

Have a client where this would help, we are using validation functions to fix a bug with regular online orders however this same validation function is now causing issue and prevent draft orders from working

@Oteiza-a not sure if be the case but I do know there are some things feature flagged in the Shopify Admin, so you may need to reach out to support and ask

**Reply by gil--** [reply]
Hi @Jrschellenberg, we don't plan to provide "surface" data such as draft vs regular checkout in the function input graph. Sounds like this function was to fix a bug so I'd recommend opening a ticket or posting on https://community.shopify.dev/c/shopify-functions/15 and we can look into the root cause.
